src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        in the 'PlineMarkers' list)
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        wflow['IsClean'] = self._checkClean(wflow)
+        return wflow
+
+    def archive(self, wflow):
+        """
+        Move the workflow to the proper archived status after checking
+        the full cleanup status
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :param  archStatus: Target status to transition after archival
+        :return wflow:
+        """
+        # NOTE: check allowed status transitions with:
+        #       https://github.com/dmwm/WMCore/blob/5961d2229b1e548e58259c06af154f33bce36c68/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L171
+        if not wflow['IsClean']:
+            msg = "Not properly cleaned workflow: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+
+        # Check the available status transitions before we decide the final status
+        targetStatusList = RequestStatus.REQUEST_STATE_TRANSITION.get(wflow['RequestStatus'], [])
+        self.logger.info("targetStatusList: %s", targetStatusList)
+        return wflow
+
+    def getMSOutputTransferInfo(self, wflow):
+        """
+        Fetches the transfer information from the MSOutput REST interface for
+        the given workflow.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        headers = {'Accept': 'application/json'}
+        params = {}
+        url = '%s/data/info?request=%s' % (self.msConfig['MSOutputUrl'],
+                                           wflow['RequestName'])
+        mgr = RequestHandler()
+        res = mgr.getdata(url, params=params, headers=headers, ckey=ckey(), cert=cert())
+        data = json.loads(res)['result'][0]
+        wflow['TransferInfo'] = data['transferDoc']
+        return wflow
+
+    def getRucioRules(self, wflow, gran, rucioAcct):
+        """
+        Queries Rucio and builds the relevant list of blocklevel rules for
+        the given workflow
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :param  gran:    Data granularity to search for Rucio rules. Possible values:
+                        'block' || 'container'
+        :return:        wflow
+        """
+        currPline = wflow['PlineMarkers'][-1]
+        # Find all the output placement rules crated by the agents
+        # NOTE: Check MSOutput 'TransferStatus' first and assure it is 'done'
+        #       only then we proceed with building the respective list of rules
+        if wflow['TransferInfo'] is not None \
+           and wflow['TransferInfo']['TransferStatus'] == 'done': # FIXME 'pennding' -> 'done'
+            for dataCont in wflow['OutputDatasets']:
+                if gran == 'container':
+                    wflow['RulesToClean'][currPline].extend(self.rucio.listDataRules(dataCont, account=rucioAcct))
```

Comment:
I guess you want to let Rucio module exceptions to propagate upstream, right?

Reply:
Yes, indeed. The exception will be caught up in the `_dispatcher()`, where the `pline.run()` method is called and the the execution for this workflow will be terminated.